packages/react-redux-subspace/src/components/SubspaceProvider.js

```diff
@@ -25,22 +58,40 @@ class SubspaceProvider extends React.PureComponent {
     }
 }
 
+SubspaceLegacyContextOutput.childContextTypes = {
+    store: PropTypes.object
+}
+
+class SubspaceProvider extends React.Component {
+    makeSubspace = memoizeOne((mapState, namespace, subspaceDecorator, store) => {
+        const createSubspace = subspaceDecorator || subspace(mapState, namespace)
+        return createSubspace(store)
+    })
+
+    render() {
+        const { mapState, namespace, subspaceDecorator, store, children } = this.props
+        const subspace = this.makeSubspace(mapState, namespace, subspaceDecorator, store)
+        return (
+            <Provider key={subspace.namespace} store={subspace}>
+                <SubspaceLegacyContextOutput subspace={subspace}>
+                    {Children.only(children)}
+                </SubspaceLegacyContextOutput>
+            </Provider>
+        )
+    }
+}
+
 SubspaceProvider.propTypes = {
     children: PropTypes.element.isRequired,
+    store: PropTypes.object.isRequired,
     mapState: PropTypes.oneOfType([
         PropTypes.func,
         PropTypes.string,
     ]),
     namespace: PropTypes.string,
-    subspaceDecorator: PropTypes.func,
-}
-
-SubspaceProvider.contextTypes = {
-    store: PropTypes.object.isRequired
+    subspaceDecorator: PropTypes.func
 }
 
-SubspaceProvider.childContextTypes = {
-    store: PropTypes.object
-}
+const storeAccess = compose(contextApi, legacyContext)
 
-export default SubspaceProvider
+export default storeAccess(SubspaceProvider)
```

Comment:
because this maintains legacy context support, subspacing a single component using

```js
<SubspaceProvider mapState={(state) => state.example} namespace='example>
  <Component />
</SubspaceComponent>
```

ends up with a component heirarchy of

```js
<SubspaceContextApi>
  <ReactReduxContext.Consumer>
    <SubspaceLegacyContextInput>
      <SubspaceProvider>
        <Provider>
          <ReactReduxContext.Provider>
            <SubspaceLegacyContextOutput>
              <Component />
            </SubspaceLegacyContextOutput>
          </ReactReduxContext.Provider>
        </Provider>
      </SubspaceComponent>
    </SubspaceLegacyContextInput>
  </ReactReduxContext.Consumer>
</SubspaceContextApi>
```

If you add in the `subspaced` HOC, the top level `Provider` and the layers for the `connect` HOC of the component, this starts to look a bit crazy.  In fairness, the legacy context support is only 2 of the 7 layers, so not having it only makes the problem a little bit better.

Reply:
Anyone running an app complex enough to need subspace likely has worse demons in their component tree.